### PR TITLE
Basic implementation of ArrayEnumerator

### PR DESCRIPTION
**Pull request by saveriomiroddi**
Basic implementation, with `#next` and `#has_next?` methods; the new `StopIteration` error class is also included.

This class is required for implementing the Array lazy enumerator (#537).

~@st0012: there are a couple of things to straighten:~

~1. there isn't an appropriate error - all the errors are specific, and InternalError is not appropriate either; implementing a StopIteration would actually suit well, due to the upcoming LazyEnumerator implementation~
~2. there isn't any support for raising errors (`raise StandardError.new('Error message')`~

### CODE REVIEW

**Comment by st0012** [comment]
> @st0012: there are a couple of things to straighten:
> 1. there isn't an appropriate error - all the errors are specific, and InternalError is not appropriate either; implementing a StopIteration would actually suit well, due to the upcoming LazyEnumerator implementation
> 2. there isn't any support for raising errors (raise StandardError.new('Error message')

@saveriomiroddi For the first point, you can just add any new error you want.
But the second point requires other works and should take some time.

**Comment by saveriomiroddi** [comment]
@st0012 ready!

**Comment by st0012** [comment]
@saveriomiroddi well done!

**Comment by saveriomiroddi** [comment]
@st0012 thanks :smile: The LazyEnumerator is on a separate PR (#573); once the block code is merged in master, I'll finalize it.

**Comment by saveriomiroddi** [comment]
@st0012 this needs to be merged in order for the lazy enumerator to be merged as well, since I'm treating as two separate features.

File: lib/array.gb [hunk 1]
```diff
@@ -4,4 +4,8 @@ class Array
       i == x
     end
   end
+
+  def to_enum
+		ArrayEnumerator.new(self)
```
Comment (st0012) [inline lib/array.gb]: @saveriomiroddi  Can you make this indentation right? After that I think we'll be ok to merge this

**Pull request merged**